### DISCUSSION TITLE
passing opaque pointers

### DISCUSSION THREAD

**Discussion by bmahlbrand** [wjakob/nanobind#11] [Q&A] [answered]
I want to accomplish something like the following:

```
ctx = module.init()

start(ctx)
end(ctx)
```
where ctx is some opaque struct pointer that python needs not be aware of (not accessing private members or anything)

using the following compiles fine, but complains when ```ctx``` is a valid pointer returned from this signature (ie passing null pointers around works fine):
```
context_state_struct *init(std::string path) {
    ...
````
thus
```
ctx = module.init()
```
crashes with the error - `TypeError: Unable to convert function return value to a Python type! The signature was`

I looked at `nb::handle<>` and `nb::handle_of<>` but not sure if there's something else I should be doing here.

Thanks!

**Comment by wjakob** [comment, marked as answer]
You need a `nb::class_` declaration for `context_state_struct`. No need to provide a constructor/method/fields.

**Reply by bmahlbrand** [reply]
Thanks! That did it - for future travelers, I ended up with something like this:

```nb::class_<context_state_struct>(m, "context_state_struct");```
